### PR TITLE
Add overview documentation about event implementation.

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

**Review by dweindl — APPROVED** [review]
:+1:

File: documentation/implementation_events.rst [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+only track zero crossings from negative to positive. Accordingly two root
+functions are necessary to keep track of Heaviside functions and two
+Heaviside function helper variables will be created, where one corresponds
+to the value of `Heaviside(...)` and one to the value of `1-Heaviside(...)`.
+
+
+Solution Jump Discontinuities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+SUNDIALS by itself does not support solution jump discontinuities. We
```
Comment (dweindl) [inline documentation/implementation_events.rst:73]: Worth a feature request?

File: documentation/implementation_events.rst [hunk 2]
```diff
@@ -0,0 +1,85 @@
+Event Implementation
+=======================
+
+This document provides guidance an rational on the implementation of events in
+AMICI. Events include any discontinuities in the right hand side of the
+differential equation. There are three types of discontinuities:
+
+- **Solution Jump Discontinuities** can be created by SBML  events or delta
```
Comment (dilpath) [inline documentation/implementation_events.rst:8]:
```suggestion
- **Solution Jump Discontinuities** can be created by SBML events or delta
```

File: documentation/implementation_events.rst [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  functions in the right hand side.
+
+- **Right-Hand-Side Jump Discontinuities** result in removable
+  discontinuities in the solution and can be created by SBML Piecewise or
+  Heaviside functions.
+
+- **Right-Hand-Side Removeable Discontinuities** do not lead to
```
Comment (dilpath) [inline documentation/implementation_events.rst:15]:
```suggestion
- **Right-Hand-Side Removable Discontinuities** do not lead to
```

File: documentation/implementation_events.rst [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+Right-Hand-Side Jump Discontinuities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In principle these discontinuities do not need any special treatment, but
+empirically, the solver may overstep or completely ignore the discontinuity,
+leading to poor solution quality. This is particularly problematic when
+step size is large and changes in step size, which can be caused by
+parameter changes, inclusion of forward sensitivities or during backward
+solves, may alter solutions in unexpected ways. Accordingly finite
+difference approximations, forward sensitivies as well as adjoint
```
Comment (dilpath) [inline documentation/implementation_events.rst:43]:
```suggestion
difference approximations, forward sensitivities as well as adjoint
```

File: documentation/implementation_events.rst [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+sensitivies will yield poor derivative approximations.
```
Comment (dilpath) [inline documentation/implementation_events.rst:44]:
```suggestion
sensitivities will yield poor derivative approximations.
```

File: documentation/implementation_events.rst [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+To address these issues, we use the built-in rootfinding functionality in
+SUNDIALS, WHICH pauses the solver at the locations of discontinuities and
+avoids overstepping or ignoring of discontinuities.
+
+Another difficulty comes with the evaluation of heaviside functions. After
+or during processing of discontinuities, Heaviside functions need to be
+evaluated at the left and right hand limit of discontinuities.
+This is challenging as the solver may slightly over or understep the
+discontinuity timepoint by a small epsilon and limits have to be correctly
+computed in both forward and backward pass.
```
Comment (dilpath) [inline documentation/implementation_events.rst:55]:
```suggestion
computed in both forward and backward passes.
```

File: documentation/implementation_events.rst [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+Right-Hand-Side Jump Discontinuities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In principle these discontinuities do not need any special treatment, but
+empirically, the solver may overstep or completely ignore the discontinuity,
+leading to poor solution quality. This is particularly problematic when
+step size is large and changes in step size, which can be caused by
+parameter changes, inclusion of forward sensitivities or during backward
+solves, may alter solutions in unexpected ways. Accordingly finite
```
Comment (dilpath) [inline documentation/implementation_events.rst:42]:
```suggestion
solves, may alter solutions in unexpected ways. Accordingly, finite
```

File: documentation/implementation_events.rst [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+SUNDIALS by itself does not support solution jump discontinuities. We
+implement support by accessing private SUNDIALS API in
+:cpp:func:`Solver::resetState`, :cpp:func:`Solver::reInitPostProcess` and
+:cpp:func:`Solver::reInitPostProcessB`. These functions resets interval
+variables to initial values to simulate a fresh integration start, but
+keep/update the solution history, which is important for adjoint solutions.
+
+
+Right-Hand-Side Removeable Discontinuities
```
Comment (dilpath) [inline documentation/implementation_events.rst:81]:
```suggestion
Right-Hand-Side Removable Discontinuities
```

File: documentation/implementation_events.rst [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Removeable discontinuities do not require any special treatment. Numerically
```
Comment (dilpath) [inline documentation/implementation_events.rst:84]:
```suggestion
Removable discontinuities do not require any special treatment. Numerically,
```

File: documentation/implementation_events.rst [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+* Fröhlich, F., Theis, F. J., Rädler, J. O., & Hasenauer, J. (2017).
+  Parameter estimation for dynamical systems with discrete events and logical
+  operations. Bioinformatics, 33(7), 1049-1056.
+  doi:`10.1093/bioinformatics/btw764 <https://doi.org/10.1093/bioinformatics/btw764>`_.
+
+Algorithmic Considerations
+--------------------------
+
+Right-Hand-Side Jump Discontinuities
```
Comment (dilpath) [inline documentation/implementation_events.rst:34]: Order of subsections in "Algorithmic Considerations" is different to the order of items in the bold list above.

**Review by dilpath — APPROVED** [review]
Seems SBML-centric, so the title could be changed to "SBML Event Implementation". Thanks for the write-up!

**Pull request merged**